File: src/main/java/com/chatchatabc/jpademojava/impl/domain/service/PostServiceImpl.java
package com.chatchatabc.jpademojava.impl.domain.service;

import com.chatchatabc.jpademojava.domain.model.Post;
import com.chatchatabc.jpademojava.domain.model.User;
import com.chatchatabc.jpademojava.domain.repository.PostRepository;
import com.chatchatabc.jpademojava.domain.repository.UserRepository;
import com.chatchatabc.jpademojava.domain.service.PostService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class PostServiceImpl implements PostService {
    @Autowired
    private PostRepository postRepository;
    @Autowired
    private UserRepository userRepository;

    /**
     * Create post
     *
     * @param userId
     * @param post
     * @return
     */
    @Override
    public Post create(String userId, Post post) throws Exception {
        Optional<User> user = userRepository.findById(userId);
        if (user.isEmpty()) {
            throw new Exception("User not found");
        }
        post.setUser(user.get());
        return postRepository.save(post);
    }

    /**
     * Update posts
     *
     * @param postId
     * @param newPostInfo
     * @return
     */
    @Override
    public Post update(String postId, Post newPostInfo) throws Exception {
        Optional<Post> post = postRepository.findById(postId);
        if (post.isEmpty()) {
            throw new Exception("Post not found");
        }

        // Apply updates
        if (newPostInfo.getTitle() != null) {
            post.get().setTitle(newPostInfo.getTitle());
        }
        if (newPostInfo.getContent() != null) {
            post.get().setContent(newPostInfo.getContent());
        }
        return postRepository.save(post.get());
    }

    /**
     * Delete post
     *
     * @param postId
     */
    @Override
    public void delete(String postId) throws Exception {
        // Not really needed
        Optional<Post> post = postRepository.findById(postId);
        if (post.isEmpty()) {
            throw new Exception("Post not found");
        }
        postRepository.delete(post.get());
    }
}

File: src/main/kotlin/com/chatchatabc/jpademo/impl/domain/service/PostServiceImpl.kt
package com.chatchatabc.jpademo.impl.domain.service

import com.chatchatabc.jpademo.domain.model.Post
import com.chatchatabc.jpademo.domain.repository.PostRepository
import com.chatchatabc.jpademo.domain.repository.UserRepository
import com.chatchatabc.jpademo.domain.service.PostService
import org.springframework.stereotype.Service

@Service
class PostServiceImpl(
    private val postRepository: PostRepository,
    private val userRepository: UserRepository
) : PostService {

    /**
     * Create Post
     */
    override fun create(userId: String, post: Post): Post {
        val user = userRepository.findById(userId)
        if (user.isEmpty) {
            throw Exception("User not found")
        }
        post.user = user.get()
        return postRepository.save(post)
    }

    /**
     * Update Post
     */
    override fun update(postId: String, newPostInfo: Post): Post {
        val post = postRepository.findById(postId)
        if (post.isEmpty) {
            throw Exception("Post not found")
        }
        if (newPostInfo.title != null) {
            post.get().title = newPostInfo.title
        }
        if (newPostInfo.content != null) {
            post.get().content = newPostInfo.content
        }
        return postRepository.save(post.get())
    }

    /**
     * Delete Post
     */
    override fun delete(postId: String) {
        // Not really needed
        val post = postRepository.findById(postId)
        if (post.isEmpty) {
            throw Exception("Post not found")
        }
        postRepository.delete(post.get())
    }
}